doc/introduction/compiling_circuits.rst

```diff
@@ -0,0 +1,252 @@
+
+    Most compilation transforms support just-in-time compilation with ``jax.jit``.
+
+The :func:`~.pennylane.compile` transform allows you to chain together
+sequences of quantum function transforms into custom circuit optimization pipelines.
+
+For example, take the following decorated quantum function:
+
+.. code-block:: python
+
+    dev = qml.device('default.qubit', wires=[0, 1, 2])
+
+    @qml.qnode(dev)
+    @qml.compile()
+    def qfunc(x, y, z):
+        qml.Hadamard(wires=0)
+        qml.Hadamard(wires=1)
+        qml.Hadamard(wires=2)
+        qml.RZ(z, wires=2)
+        qml.CNOT(wires=[2, 1])
+        qml.RX(z, wires=0)
+        qml.CNOT(wires=[1, 0])
+        qml.RX(x, wires=0)
+        qml.CNOT(wires=[1, 0])
+        qml.RZ(-z, wires=2)
+        qml.RX(y, wires=2)
+        qml.PauliY(wires=2)
+        qml.CZ(wires=[1, 2])
+        return qml.expval(qml.PauliZ(wires=0))
+
+The default behaviour of :func:`~.pennylane.compile` is to apply a sequence of three
+transforms: :func:`~.pennylane.transforms.commute_controlled`, :func:`~.pennylane.transforms.cancel_inverses`,
+and then :func:`~.pennylane.transforms.merge_rotations`.
+
+>>> print(qml.draw(qfunc)(0.2, 0.3, 0.4))
+0: ──H───RX(0.6)──────────────────┤ ⟨Z⟩
+1: ──H──╭X────────────────────╭C──┤
+2: ──H──╰C────────RX(0.3)──Y──╰Z──┤
+
+
+The :func:`~.pennylane.compile` transform is flexible and accepts a custom pipeline
+of quantum function transforms (you can even write your own!).
+For example, if we wanted to only push single-qubit gates through
+controlled gates and cancel adjacent inverses, we could do:
+
+.. code-block:: python
+
+    from pennylane.transforms import commute_controlled, cancel_inverses
+    pipeline = [commute_controlled, cancel_inverses]
+
+    @qml.qnode(dev)
+    @qml.compile(pipeline=pipeline)
+    def qfunc(x, y, z):
+        qml.Hadamard(wires=0)
+        qml.Hadamard(wires=1)
+        qml.Hadamard(wires=2)
+        qml.RZ(z, wires=2)
+        qml.CNOT(wires=[2, 1])
+        qml.RX(z, wires=0)
+        qml.CNOT(wires=[1, 0])
+        qml.RX(x, wires=0)
+        qml.CNOT(wires=[1, 0])
+        qml.RZ(-z, wires=2)
+        qml.RX(y, wires=2)
+        qml.PauliY(wires=2)
+        qml.CZ(wires=[1, 2])
+        return qml.expval(qml.PauliZ(wires=0))
+
+>>> print(qml.draw(qfunc)(0.2, 0.3, 0.4))
+0: ──H───RX(0.4)──RX(0.2)────────────────────────────┤ ⟨Z⟩
+1: ──H──╭X───────────────────────────────────────╭C──┤
+2: ──H──╰C────────RZ(0.4)──RZ(-0.4)──RX(0.3)──Y──╰Z──┤
+
+.. note::
+
+    The :class:`~.pennylane.Barrier` operator can be used to prevent blocks of code from being merged during
+    compilation.
+
+
+For more details on :func:`~.pennylane.compile` and the available compilation transforms, visit
+`the compilation documentation
+<../code/qml_transforms.html#transforms-for-circuit-compilation>`_.
+
+Groups of commuting Pauli words
+-------------------------------
+
+Mutually commuting Pauli words can be measured simultaneously on a quantum computer.
```

Comment:
If that changes soon, I wonder if we should leave the ambiguity here, and people can refer to the docstring for details? I fear we forget to update...

Reply:
And I think it's already been merged in... So leave it be.

Reply:
@mariaschuld I might suggest having:

```rest
.. autosummary::

  ~split_non_commuting
  ~hamiltonian_expand
  ~group_observables
```

here? 

Split non-commuting is this same functionality, but at the QNode level:

```python
@qml.transforms.split_non_commuting
@qml.qnode(dev)
def circuit(x):
    qml.RX(x,wires=0)
    return [qml.expval(qml.PauliX(0)), qml.expval(qml.PauliZ(0))]

>>> print(qml.draw(circuit)(0.5))
0: ──RX(0.50)─┤  <X>
0: ──RX(0.50)─┤  <Z>
```

Hamiltonian expand is the same but for Hamiltonians. It actually _should_ support QNodes! This is just an oversight, that can be fixed by adding `@qml.batch_transform` above it's definition lol:

```python
# we should do this in the library!
hamiltonian_expand = qml.batch_transform(qml.transforms.hamiltonian_expand)

@hamiltonian_expand
@qml.qnode(dev)
def circuit(x):
    qml.RX(x,wires=0)
    return qml.expval(qml.PauliX(0) + qml.PauliZ(0))

>>> print(qml.draw(circuit)(0.5))
0: ──RX(0.50)─┤  <X>
0: ──RX(0.50)─┤  <Z>
```

I think only then would I introduce `group_observables`, for those working at a low level. Because it isn't clear to me from reading this paragraph _how to use_ `group_observables` with an existing Hamiltonian or QNode.

In fact, it might be better to first describe `H.compute_grouping()` _first_?